iml-gui/crate/src/components/command_modal.rs

```diff
@@ -284,3 +547,643 @@ fn close_button() -> Node<Msg> {
+            state: "complete".to_string(),
+            step_results: Default::default(),
+            steps: steps.iter().map(|x| format!("/api/step/{}/", x)).collect(),
+            wait_for: wait_for.iter().map(|x| format!("/api/job/{}/", x)).collect(),
+            write_locks: vec![],
+        }
+    }
+
+    fn make_step(id: u32, class_name: &str) -> Step {
+        Step {
+            args: Default::default(),
+            backtrace: "".to_string(),
+            class_name: class_name.to_string(),
+            console: "".to_string(),
+            created_at: "2020-03-16T07:22:34.491600".to_string(),
+            description: "".to_string(),
+            id,
+            log: "".to_string(),
+            modified_at: "2020-03-16T07:22:34.491600".to_string(),
+            resource_uri: format!("/api/step/{}/", id),
+            result: None,
+            state: "incomplete".to_string(),
+            step_count: 0,
+            step_index: 0,
+        }
+    }
+
+    fn build_db() -> Db {
+        let all_cmds = vec![
+            make_command(1, &[10, 11], "One"),
+            make_command(2, &[12, 13], "Two"),
+            make_command(3, &[14, 15], "Three"),
+            make_command(4, &[16, 17], "Four"),
+        ];
+        let all_jobs = vec![
+            make_job(10, &[20, 21], &[11], "Ten"),
+            make_job(11, &[21, 26], &[], "Eleven"),
+            make_job(12, &[22, 23], &[13], "Twelve"),
+            make_job(13, &[23, 28], &[], "Thirteen"),
+            make_job(14, &[24, 15], &[], "Ten"),
+            make_job(15, &[25, 20], &[], "Eleven"),
+            make_job(16, &[26, 27], &[], "Twelve"),
+            make_job(17, &[27, 22], &[], "Thirteen"),
+        ];
+        let all_steps = vec![
+            make_step(20, "Twenty and zero"),
+            make_step(21, "Twenty and one"),
+            make_step(22, "Twenty and two"),
+            make_step(23, "Twenty and three"),
+            make_step(24, "Twenty and four"),
+            make_step(25, "Twenty and five"),
+            make_step(26, "Twenty and six"),
+            make_step(27, "Twenty and seven"),
+            make_step(28, "Twenty and eight"),
+            make_step(29, "Twenty and nine"),
+        ];
+        Db {
+            all_cmds,
+            all_jobs,
+            all_steps,
+        }
+    }
+
+    fn prepare_subset(db: &Db, id: u32) -> (Vec<Command>, Vec<Job0>, Vec<Step>) {
+        let cmds = db.select_cmds(&vec![id]);
+        let c_ids = cmds
+            .iter()
+            .map(|x| extract_ids::<Job0>(&x.jobs))
+            .flatten()
+            .collect::<Vec<u32>>();
+        let jobs = db.select_jobs(&c_ids);
+        let j_ids = jobs
+            .iter()
+            .map(|x| extract_ids::<Step>(&x.steps))
+            .flatten()
+            .collect::<Vec<u32>>();
+        let steps = db.select_steps(&j_ids);
+        let cmds = db.all_cmds.clone(); // use all roots
+        (cmds, jobs, steps)
+    }
```

Comment:
```suggestion
  fn prepare_subset(db: &Db, id: u32) -> (Vec<&Command>, Vec<&Job0>, Vec<&Step>) {
        let cmds = db.select_cmds(&vec![id]);
        let c_ids = cmds
            .iter()
            .map(|x| extract_ids::<Job0>(&x.jobs))
            .flatten()
            .collect::<Vec<u32>>();
        let jobs = db.select_jobs(&c_ids);
        let j_ids = jobs
            .iter()
            .map(|x| extract_ids::<Step>(&x.steps))
            .flatten()
            .collect::<Vec<u32>>();
        let steps = db.select_steps(&j_ids);
        let cmds = db.all_cmds.iter().collect(); // use all roots
        (cmds, jobs, steps)
    }
```

No need to clone here.

Reply:
Sorry, Joe, not that easy. :mask:

The thing is that `prepare_subset` function resembles going for the data from the remote API. The handlers `Msg::FetchedCommands`, `Msg::FetchedJobs` and `Msg::FetchedSteps` accept `ApiList<T>` and they need to build vectors of `RichCommand`s, `RichJob`s and `RichStep`s respectively (therefore they need to own the contents!). As a consequence, `assign_commands` cannot accept __something of__ `&Command`. The same is for jobs and steps.

I removed the unnecessary clones in the `assign_commands`, `assign_jobs` and `assign_steps` by changing their signatures:
```rust
+    fn assign_commands(&mut self, cmds: Vec<Command>) {
+        self.commands = convert_to_rich_hashmap(cmds, extract_children_from_cmd);
-     fn assign_commands(&mut self, cmds: &[Command]) {
-        self.commands = convert_to_rich_hashmap(cmds.to_vec(), extract_children_from_cmd);
```
but other than that I cannot see an easy way of getting rid of `clone`s.

I could potentially get rid of `Rich<K, V>` wrapper, also reducing the number of clones needed, however then I needed to create an `Iterator<Item=u32>` over the vector of strings like `vec!["/api/command/111/", "/api/command/112/", "/api/command/113/"]`, and this parsing was happen each time the method `.deps()` is accessed.

Reply:
Hrrm, I tried this locally yesterday and everything compiled except the tests, which I did not bother to update.